ttnn/cpp/ttnn/operations/moreh/moreh_clip_grad_norm/moreh_clip_grad_norm_step1/device/moreh_clip_grad_norm_step1_program_factory.cpp

```diff
@@ -0,0 +1,224 @@
+//
+// SPDX-License-Identifier: Apache-2.0
+
+#include <vector>
+
+#include "moreh_clip_grad_norm_step1_device_operation.hpp"
+#include "tt_metal/common/work_split.hpp"
+#include "ttnn/operations/core/compute_kernel/compute_kernel_config.hpp"
+#include "ttnn/operations/moreh/moreh_helper_functions.hpp"
+
+namespace ttnn::operations::moreh::moreh_clip_grad_norm_step1 {
+
+std::tuple<uint32_t, float, bool> get_p_decimal_p_is_negative(float ord) {
+    auto p = std::floor(ord);
+    auto decimal = ord - p;
+    const bool p_is_negative = p < 0.0f;
+    if (p_is_negative) {
+        p = -p;
+    }
+    return std::make_tuple(static_cast<uint32_t>(p), decimal, p_is_negative);
+}
+
+MorehClipGradNormStep1Operation::ProgramFactory::cached_program_t
+MorehClipGradNormStep1Operation::ProgramFactory::create(
+    const operation_attributes_t& operation_attributes,
+    const tensor_args_t& tensor_args,
+    tensor_return_value_t& tmp_pow_sum) {
+    auto& inputs = tensor_args.inputs;
+    auto norm_type = operation_attributes.norm_type;
+    auto tile_offset_of_tmp_pow_sum = operation_attributes.tile_offset_of_tmp_pow_sum;
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                      Device Setup
+    ////////////////////////////////////////////////////////////////////////////
+    auto device = tmp_pow_sum.device();
+    auto program = CreateProgram();
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                         Parameters Setup
+    ////////////////////////////////////////////////////////////////////////////
+    const auto num_inputs = static_cast<uint32_t>(inputs.size());
+
+    std::vector<std::pair<uint32_t, uint32_t>> origin_hw_vec;
+    origin_hw_vec.reserve(num_inputs);
+
+    for (uint32_t j = 0; j < num_inputs; ++j) {
+        // TODO: Duong
+        const auto& input_shape_without_padding = inputs.at(j).get_logical_shape();
+        // const auto& input_shape_without_padding = inputs.at(j).get_legacy_shape().without_padding();
+        origin_hw_vec.emplace_back(input_shape_without_padding[2], input_shape_without_padding[3]);
+    }
+
+    auto [p, decimal, p_is_negative] = get_p_decimal_p_is_negative(norm_type);
+
+    ////////////////////////////////////////////////////////////////////////////
+    //                         Core Setup
+    ////////////////////////////////////////////////////////////////////////////
+    auto grid = device->compute_with_storage_grid_size();
+    const auto num_cores_x = grid.x;
+    const auto num_cores_y = grid.y;
+    const auto
+        [num_cores_to_be_used,
+         all_cores,
+         core_group_1,
+         core_group_2,
+         num_inputs_per_core_group_1,
+         num_inputs_per_core_group_2] = tt::tt_metal::split_work_to_cores(grid, num_inputs);
+    TT_ASSERT(core_group_2.ranges().empty());
+    TT_ASSERT(num_inputs_per_core_group_1 == 1);
+    TT_ASSERT(num_inputs_per_core_group_2 == 0);
```

Comment:
```suggestion
    TT_FATAL(core_group_2.ranges().empty());
    TT_FATAL(num_inputs_per_core_group_1 == 1);
    TT_FATAL(num_inputs_per_core_group_2 == 0);
```

Reply:
Resolved in [06f7fe7](https://github.com/tenstorrent/tt-metal/pull/14667/commits/06f7fe78b8714d808e3442033a18a39ce88500c8)